src/Columns/IColumn.h

```diff
@@ -49,6 +49,40 @@ struct EqualRange
 
 using EqualRanges = std::vector<EqualRange>;
 
+/// A checkpoint that contains size of column and all its subcolumns.
+/// It can be used to rollback column to the previous state, for example
+/// after failed parsing when column may be in inconsistent state.
+struct ColumnCheckpoint
+{
+    explicit ColumnCheckpoint(size_t size_) : size(size_) {}
+    size_t size = 0;
+};
+
+using ColumnCheckpointPtr = std::shared_ptr<const ColumnCheckpoint>;
+using ColumnCheckpoints = std::vector<ColumnCheckpointPtr>;
+
+struct ColumnCheckpointWithNested : public ColumnCheckpoint
+{
+    ColumnCheckpointWithNested(size_t size_, ColumnCheckpointPtr nested_)
+        : ColumnCheckpoint(size_)
+        , nested(std::move(nested_))
+    {
+    }
+
+    ColumnCheckpointPtr nested;
+};
+
+struct ColumnCheckpointWithNestedTuple : public ColumnCheckpoint
```

Comment:
Maybe use more generic name as it's used not only in `Tuple` type? Like `ColumnCheckpointWithMultipleNested`